src/acc/comms_kernel.cpp

```diff
@@ -0,0 +1,266 @@
+    //				globals.chunk.right_snd);
+    //		globals.chunk.hm_right_rcv = Kokkos::create_mirror_view(
+    //				globals.chunk.right_rcv);
+    //		globals.chunk.hm_bottom_snd = Kokkos::create_mirror_view(
+    //				globals.chunk.bottom_snd);
+    //		globals.chunk.hm_bottom_rcv = Kokkos::create_mirror_view(
+    //				globals.chunk.bottom_rcv);
+    //		globals.chunk.hm_top_snd = Kokkos::create_mirror_view(globals.chunk.top_snd);
+    //		globals.chunk.hm_top_rcv = Kokkos::create_mirror_view(globals.chunk.top_rcv);
+  }
+}
+
+void clover_exchange(global_variables &globals, const int fields[NUM_FIELDS], const int depth) {
+
+  // Assuming 1 patch per task, this will be changed
+
+  int left_right_offset[NUM_FIELDS];
+  int bottom_top_offset[NUM_FIELDS];
+
+  MPI_Request request[4] = {0};
+  int message_count = 0;
+
+  int end_pack_index_left_right = 0;
+  int end_pack_index_bottom_top = 0;
+  for (int field = 0; field < NUM_FIELDS; ++field) {
+    if (fields[field] == 1) {
+      left_right_offset[field] = end_pack_index_left_right;
+      bottom_top_offset[field] = end_pack_index_bottom_top;
+      end_pack_index_left_right += depth * (globals.chunk.y_max + 5);
+      end_pack_index_bottom_top += depth * (globals.chunk.x_max + 5);
+    }
+  }
+
+  static clover::Buffer1D<double> left_rcv_buffer(globals.context, end_pack_index_left_right);
+  static clover::Buffer1D<double> left_snd_buffer(globals.context, end_pack_index_left_right);
+  static clover::Buffer1D<double> right_rcv_buffer(globals.context, end_pack_index_left_right);
+  static clover::Buffer1D<double> right_snd_buffer(globals.context, end_pack_index_left_right);
+
+  static clover::Buffer1D<double> top_rcv_buffer(globals.context, end_pack_index_bottom_top);
+  static clover::Buffer1D<double> top_snd_buffer(globals.context, end_pack_index_bottom_top);
+  static clover::Buffer1D<double> bottom_rcv_buffer(globals.context, end_pack_index_bottom_top);
+  static clover::Buffer1D<double> bottom_snd_buffer(globals.context, end_pack_index_bottom_top);
+
+  double *left_rcv = left_rcv_buffer.data;
+  double *left_snd = left_snd_buffer.data;
+  double *right_rcv = right_rcv_buffer.data;
+  double *right_snd = right_snd_buffer.data;
+  double *top_rcv = top_rcv_buffer.data;
+  double *top_snd = top_snd_buffer.data;
+  double *bottom_rcv = bottom_rcv_buffer.data;
+  double *bottom_snd = bottom_snd_buffer.data;
+
+  bool stage = globals.config.staging_buffer;
+
+#pragma acc data create(left_rcv[ : left_rcv_buffer.N()]) create(left_snd[ : left_snd_buffer.N()])                  \
+    create(right_rcv[ : right_rcv_buffer.N()]) create(right_snd[ : right_snd_buffer.N()])                                        \
+    create(top_rcv[ : top_rcv_buffer.N()]) create(top_snd[ : top_snd_buffer.N()])                                                \
+    create(bottom_rcv[ : bottom_rcv_buffer.N()]) create(bottom_snd[ : bottom_snd_buffer.N()])
+  {
```

Comment:
We have a bug here if we 1) use a CUDA-aware MPI or 2) specify `--staging true`  in the CLI and have >1 MPI ranks.
On NVHPC 23.9 (CUDA 12.2) with the bundled HPC-X MPI, there's a ~35% deviation from the solution on `clover_bm16_short.in`.

I'm not too well versed in OpenACC, but to pass validation, I made the following changes:

1. Replace this directive with `#pragma acc enter data`, drop the extra `{}` (`acc exit data` stays the same).
2. Replace all uses of `#pragma acc data if (!stage) deviceptr(...)` with `#pragma acc host_data if (!stage) use_device(...)`, as we're doing an CUDA-aware MPI call in the block, not a device-side kernel.

Can you double check whether you can reproduce this?

Reply:
Are you sure you get a ~35% deviation when you specify `--staging-buffer true`? I get that ~35 deviation when I do a `--staging-buffer false`. I have attached my [output](https://github.com/UoB-HPC/CloverLeaf/files/13233350/output.txt).

Otherwise your changes did indeed pass validation for both `true` and `false` for `--staging-buffer` on `clover_bm16_short.in`.

Reply:
Sorry, it was a typo, `--staging-buffer true` forces device-to-host staging so it's the safe path.